### PR TITLE
Add FLAG Callbacks and unit test

### PR DESCRIPTION
**Pull request by Jamil**

### CODE REVIEW

File: test/contrib/flag/test_flag.py [hunk 1]
```diff
@@ -107,3 +108,58 @@ def test_zero_grad(model: torch.nn.Module, optimizer: torch.nn.Module,
...
+  def __init__(self):
+    self.counts = defaultdict(int)
+
+  def on_ascent_step_begin(self, step_index, loss):
+    self.counts['on_ascent_step_begin'] += 1
+
+  def on_ascent_step_end(self, step_index, loss, perturb_data):
+    self.counts['on_ascent_step_end'] += 1
+
+  def on_optimizer_step_begin(self, loss):
```
Comment (tnichols89) [inline test/contrib/flag/test_flag.py]: Might it be worthwhile to pass in the optimizer itself here? Same question for `on_optimizer_step_end`.

**Comment by tnichols89** [comment]
LGTM pending the one question above.

**Comment by Jamil** [comment]
Sounds good! @reislerjul do you have any suggestions on other things to pass in to the callbacks? or any additional callbacks needed?

File: torch_geometric/contrib/flag/flag.py [hunk 2]
```diff
@@ -81,15 +84,21 @@ def forward(
...
 
         # loss /= args.m
         self.loss /= n_ascent_steps
 
+        [c.on_ascent_step_end(0, self.loss, self.perturb) for c in self.callbacks]
+
         # for _ in range(args.m - 1):
-        for _ in range(n_ascent_steps - 1):
+        for i in range(n_ascent_steps - 1):
+            [c.on_ascent_step_begin(i+1, self.loss) for c in self.callbacks]
```
Comment (reislerjul) [inline torch_geometric/contrib/flag/flag.py]: super nit: `i + 1` instead of `i+1`. I'm surprised flake8 didn't complain at you!

Comment (Jamil) [reply torch_geometric/contrib/flag/flag.py]: haha ty!

File: torch_geometric/contrib/flag/flag.py [hunk 3]
```diff
@@ -2,6 +2,7 @@
 
 import torch
 
+from .flag_callback import FLAGCallback
```
Comment (reislerjul) [inline torch_geometric/contrib/flag/flag.py:5]: @Jamil I think you forgot to commit `FLAGCallback`

Comment (Jamil) [reply torch_geometric/contrib/flag/flag.py:5]: oops! the `git add -u` muscle memory is too strong

File: test/contrib/flag/test_flag.py [hunk 4]
```diff
@@ -107,3 +108,58 @@ def test_zero_grad(model: torch.nn.Module, optimizer: torch.nn.Module,
         for p in group['params']:
             assert torch.count_nonzero(
                 p.grad) == 0, "After running FLAG, the gradients should be 0!"
+
+
+class FLAGLossHistoryCallback(FLAGCallback):
```
Comment (reislerjul) [inline test/contrib/flag/test_flag.py]: not a blocker, but maybe this should be moved to `flag_callback.py` so that someone looking at the callback class has an example of the types of callbacks they can create.

File: test/contrib/flag/test_flag.py (hunk 1 above)
Comment (Jamil) [reply test/contrib/flag/test_flag.py]: on second thought, I think the `optimizer` is an object that will have the same reference throughout the lifecycle — so I'm not sure whether it actually will change on each step. It might be better to refer to `flag.optimizer` in the callback.

File: test/contrib/flag/test_flag.py (hunk 4 above)
Comment (Jamil) [reply test/contrib/flag/test_flag.py]: I was thinking about that too, I can see an argument for `FLAGLossHistoryCallback` being moved to that file, but not the counter one — that feels more like a test template rather than something that should be included in the actual codebase

Comment (reislerjul) [reply test/contrib/flag/test_flag.py]: Agreed 100%

**Comment by reislerjul** [comment]
> Sounds good! @reislerjul do you have any suggestions on other things to pass in to the callbacks? or any additional callbacks needed?

@Jamil I love the loss one! May also be helpful to have a callback that captures `perturb`.

File: test/contrib/flag/test_flag.py [hunk 5]
```diff
@@ -107,3 +108,58 @@ def test_zero_grad(model: torch.nn.Module, optimizer: torch.nn.Module,
...
+  def __init__(self):
+    self.loss_history = []
+
+  def on_ascent_step_end(self, step_index, loss, perturb):
```
Comment (reislerjul) [inline test/contrib/flag/test_flag.py]: nit: the argument here is `perturb`, but the argument in `FLAGCallback` is `perturb_data`.

**Review by reislerjul — APPROVED** [review]
This is good to go once you've made the changes you brought up yesterday :)
